src/theory/bv/bvgauss.cpp

```diff
@@ -0,0 +1,692 @@
+        }
+
+        case kind::BITVECTOR_EXTRACT:
+        {
+          unsigned w = utils::getSize(n);
+          visited[n] = w >= visited[n[0]] ? visited[n[0]] : w;
+          Assert(visited[n] <= visited[n[0]]);
+          break;
+        }
+
+        case kind::BITVECTOR_ZERO_EXTEND:
+        {
+          visited[n] = visited[n[0]];
+          break;
+        }
+
+        case kind::BITVECTOR_MULT:
+        {
+          unsigned w = 0;
+          for (unsigned i = 0, nc = n.getNumChildren(); i < nc; ++i)
+            w += visited[n[i]];
+          if (w > utils::getSize(n)) return 0; /* overflow */
+          visited[n] = w;
+          break;
+        }
+
+        case kind::BITVECTOR_CONCAT:
+        {
+          unsigned w;
+          Node zero = utils::mkZero(utils::getSize(n[0]));
+          if (n[0] == zero)
+          {
+            w = visited[n[1]]
+                /* add bw of children n[2] to n[n.getNumChildren()-1] */
+                + utils::getSize(n) - utils::getSize(n[0])
+                - utils::getSize(n[1]);
+          }
+          else
+          {
+            w = visited[n[0]]
+              /* add bw of children n[1] to n[n.getNumChildren()-1] */
+              +  utils::getSize(n) - utils::getSize(n[0]);
+          }
+          visited[n] = w;
+          break;
+        }
+
+        case kind::BITVECTOR_UDIV_TOTAL:
+        case kind::BITVECTOR_OR:
+        case kind::BITVECTOR_NOR:
+        case kind::BITVECTOR_XOR:
+        case kind::BITVECTOR_XNOR:
+        case kind::BITVECTOR_AND:
+        case kind::BITVECTOR_NAND:
+        case kind::BITVECTOR_SHL:
+        case kind::BITVECTOR_LSHR:
+        case kind::BITVECTOR_ASHR:
+        {
+          unsigned wmax = 0;
+          for (unsigned i = 0, nc = n.getNumChildren(); i < nc; ++i)
+            if (visited[n[i]] > wmax) wmax = visited[n[i]];
+          visited[n] = wmax;
+          break;
+        }
+
+        case kind::BITVECTOR_UREM_TOTAL:
+        {
+          visited[n] = visited[n[1]];
+          break;
+        }
+
+        case kind::BITVECTOR_PLUS:
+        case kind::BITVECTOR_SUB:
+        {
+          unsigned w = 0, wmax = visited[n[0]];
+          for (unsigned i = 1, nc = n.getNumChildren(); i < nc; ++i)
+            if (visited[n[i]] > wmax) wmax = visited[n[i]];
+          for (unsigned i = 0, nc = n.getNumChildren(); i < nc; ++i)
+            if (visited[n[i]] >= wmax) w += 1;
+          w += wmax - 1;
+          if (w > utils::getSize(n)) return 0; /* overflow */
+          visited[n] = w;
+          break;
+        }
+
+        case kind::BITVECTOR_NOT:
+        case kind::BITVECTOR_NEG:
+        {
+          visited[n] = visited[n[0]];
+          break;
+        }
+
+        default:
+        {
+          visited[n] = utils::getSize(n);
+        }
+      }
+    }
+  }
+  return visited[expr];
+}
+
+BVGaussElim::Result BVGaussElim::gaussElim(Integer prime,
+                                           const vector<Integer> &rhs,
```

Comment:
Random thought: It is not strictly necessary to separate the rhs from the lhs, right? It could just be kept in the same matrix, simplifying some of the code. Not really necessary, just a thought.

Reply:
Sure, this is not strictly necessary. However, using an explicit rhs IMO makes it easier to follow the algorithm (rather than having rhs implicitly as the last column of the matrix).